sw/host/opentitanlib/src/io/gpio.rs

```diff
@@ -200,3 +209,36 @@ pub trait GpioMonitoring {
         continue_monitoring: bool,
     ) -> Result<MonitoringReadResponse>;
 }
+
+/// Represents one entry in the specification of a bitbanging waveform.
+pub enum BitbangEntry<'rd, 'wr> {
+    /// Represents a sequence of pin values.  Bit 0 in each byte controls the first `GpioPin` in
+    /// the slice given to `GpioBitbanging::run()`, bit 1 controls the second `GpioPin`, and so
+    /// on.  Each byte is applied to the pins in sequence, with a particular delay between each
+    /// given by the `clock_tick` argument to `run()`.
+    Write(&'wr [u8]),
+    /// Represents a sequence of pin values as above, but additionally captures the value of any
+    /// pins in `Input` or `OpenDrain` mode.  At each clock tick, the input levels are sampled
+    /// just before the given output levels are applied, meaning that the data coming back will be
+    /// offset by one sample.
+    Both(&'wr [u8], &'rd mut [u8]),
+    /// Represents a delay of the given number of clock ticks in which the output levels are held
+    /// as indicated by the last byte of the preceeding `Write`/`Both` entry.
+    ///
+    /// A delay shorter than one tick is invalid.  A delay of one tick is equivalent to not
+    /// specifying any `Delay` between two `Write` blocks, which is also equivalent to
+    /// concatenating the two into a single `Write` block.
+    Delay(u32),
+}
+
+/// A trait implemented by transports which support synchronous bit-banging on GPIO pins, similar
+/// to FTDI devices.  This trait allows generation of arbitraty waveforms on a set of pins, and
+/// optionally getting back samples from same or other pins, taken at precise times.
+pub trait GpioBitbanging {
+    fn run(
+        &self,
+        pins: &[&dyn GpioPin],
+        clock_tick: Duration,
```

Comment:
I assume that there are some constraints on the clock ticks. Will the function return an error if this is not supported? Is there some way to report the resolution? (like `ClockNature` for GPIO monitoring)

Reply:
The Duration will be rounded to the nearest multiple of 18.1818... ns, which is the current base frequency of timers in HyperDebug (55MHz).  There is a minimum `clock_tick` duration, which I have not exactly established, and is subject to further optimizations in HyperDebug firmware.  HyperDebug is not hard realtime, so it is also very possible that there could be occasional skipped clock ticks, in case HyperDebug is simultaneously doing USB, I2C or SPI communication.
I have not made any attempt, yet, at making HyperDebug or its opentitanlib driver reject the request, if the clock period is too short.